extra/mariadb_upgrade_config_file.c

```diff
@@ -0,0 +1,1171 @@
+static void usage()
+{
+  print_version();
+  puts("This software comes with ABSOLUTELY NO WARRANTY. This is free software,\nand you are welcome to modify and redistribute it under the GPL license\n");
+  puts("Displays the unrecognized options present in configuration files, which is useful when upgrading MariaDB");
+  printf("Usage: %s [OPTIONS]\n", my_progname);
+  my_print_help(my_long_options);
+  my_print_default_files(config_file);
+  cleanup_and_exit(0);
+}
+
+
+static my_bool
+get_one_option(const struct my_option *opt __attribute__((unused)),
+	       const char *argument __attribute__((unused)),
+               const char *filename __attribute__((unused)))
+{
+  switch (opt->id) {
+    case 'I':
+    case '?':
+      usage();
+    case 'V':
+      print_version();
+      cleanup_and_exit(0);
+    case OPT_UPDATE:
+      opt_update= TRUE;
+      break;
+    case OPT_BACKUP:
+      opt_backup= TRUE;
+      break;
+    case OPT_PRINT:
+      opt_print= TRUE;
+      break;
+  }
+  return 0;
+}
+
+
+static int get_options(int *argc,char ***argv)
+{
+  int ho_error;
+
+  if ((ho_error=handle_options(argc, argv, my_long_options, get_one_option)))
+    exit(ho_error);
+
+  if (opt_update && opt_edit_mode == EDIT_MODE_NONE)
+  {
+    fputs("error: --update provided without --edit=<mode>\n", stderr);
+    exit(1);
+  }
+  if (opt_backup && !opt_update)
+  {
+    fputs("error: --backup provided without --update\n", stderr);
+    exit(1);
+  }
+  if (!opt_current_version && (opt_edit_mode == EDIT_MODE_INLINE_OLD_VERSION ||
+                               opt_edit_mode == EDIT_MODE_LAST_OLD_VERSION))
+  {
+    fputs("error: Selected --edit mode requires --current-version\n", stderr);
+    exit(1);
+  }
+  if (opt_current_version && (opt_edit_mode != EDIT_MODE_INLINE_OLD_VERSION &&
+                              opt_edit_mode != EDIT_MODE_LAST_OLD_VERSION))
+  {
+    fputs("error: --current-version provided without a corresponding --edit mode\n", stderr);
+    exit(1);
+  }
+  if (opt_print && opt_update)
+  {
+    fputs("error: --print and --update can't be specified simultaneously\n", stderr);
+    exit(1);
+  }
+
+  return 0;
+}
+
+
+int main(int argc, char **argv)
+{
+  int count, error, args_used;
+  char **load_default_groups= 0, *tmp_arguments[6];
+  char **arguments, **org_argv;
+  const char **default_directories;
+  int i= 0;
+  MY_INIT(argv[0]);
+
+  org_argv= argv;
+  args_used= get_defaults_options(argv);
+
+  /* Copy defaults-xxx arguments & program name */
+  count= args_used;
+  arguments= tmp_arguments;
+  memcpy((char*) arguments, (char*) org_argv, count*sizeof(*org_argv));
+  arguments[count]= 0;
+
+  /*
+     We already process --defaults* options at the beginning in
+     get_defaults_options(). So skip --defaults* options and
+     pass remaining options to handle_options().
+  */
+  org_argv+=args_used-1;
+  argc-=args_used-1;
+
+  /* Check out the args */
+  if (get_options(&argc,&org_argv))
+    cleanup_and_exit(1);
```

Comment:
```
CURRENT_TEST: main.mariadb_upgrade_config_file
=================================================================
==279966==ERROR: AddressSanitizer: global-buffer-overflow on address 0x55e4cf2dd2a0 at pc 0x55e4ce247595 bp 0x7ffd0749a270 sp 0x7ffd0749a268
WRITE of size 8 at 0x55e4cf2dd2a0 thread T0
    #0 0x55e4ce247594 in init_one_value /home/buildbot/amd64-ubuntu-2204-clang14-asan/build/mysys/my_getopt.c:1352:25
    #1 0x55e4ce244bc1 in init_variables /home/buildbot/amd64-ubuntu-2204-clang14-asan/build/mysys/my_getopt.c:1481:7
    #2 0x55e4ce244bc1 in handle_options /home/buildbot/amd64-ubuntu-2204-clang14-asan/build/mysys/my_getopt.c:254:5
    #3 0x55e4ce232084 in get_options /home/buildbot/amd64-ubuntu-2204-clang14-asan/build/extra/mariadb_upgrade_config_file.c:1077:17
    #4 0x55e4ce232084 in main /home/buildbot/amd64-ubuntu-2204-clang14-asan/build/extra/mariadb_upgrade_config_file.c:1139:7
    #5 0x7f3e00200d8f  (/lib/x86_64-linux-gnu/libc.so.6+0x29d8f) (BuildId: c289da5071a3399de893d2af81d6a30c62646e1e)
    #6 0x7f3e00200e3f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e3f) (BuildId: c289da5071a3399de893d2af81d6a30c62646e1e)
    #7 0x55e4ce171c34 in _start (/home/buildbot/amd64-ubuntu-2204-clang14-asan/build/extra/mariadb-upgrade-config-file+0xd4c34) (BuildId: 50912b41cfa853dccfab3aba5bc77e7ca677e090)
0x55e4cf2dd2a0 is located 32 bytes to the left of global variable 'opt_backup' defined in '/home/buildbot/amd64-ubuntu-2204-clang14-asan/build/extra/mariadb_upgrade_config_file.c' (0x55e4cf2dd2c0) of size 1
  'opt_backup' is ascii string ''
0x55e4cf2dd2a4 is located 0 bytes to the right of global variable 'opt_edit_mode' defined in '/home/buildbot/amd64-ubuntu-2204-clang14-asan/build/extra/mariadb_upgrade_config_file.c:52:23' (0x55e4cf2dd2a0) of size 4
SUMMARY: AddressSanitizer: global-buffer-overflow /home/buildbot/amd64-ubuntu-2204-clang14-asan/build/mysys/my_getopt.c:1352:25 in init_one_value
```
I don’t immediately have an idea what could be wrong here. I see that it is `my_bool opt_backup`, which should be `char` or `unsigned char`.

I was interested in the ASAN run because the [grid view for this pull request](https://buildbot.mariadb.org/#/grid?branch=refs%2Fpull%2F2976%2Fmerge) showed some memory leaks caught by `SAFEMALLOC` on a different builder. Instead of seeing leak reports, we got this in the `cmake -DWITH_ASAN=ON` build.